apis/python/src/tiledb/vector_search/object_readers/directory_reader.py

```diff
@@ -8,6 +8,223 @@
+        else:
+            return True
+    return True
+
+
+def find_uris_vfs(
+    search_uri: str,
+    config: Optional[Mapping[str, Any]] = None,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+) -> Sequence[str]:
+    with tiledb.scope_ctx(config):
+        vfs = tiledb.VFS(config=config, ctx=tiledb.Ctx(config))
+
+        def find(
+            uri: str,
+            *,
+            include: str = "*",
+            exclude: Sequence[str] = (),
+            suffixes: Optional[Sequence[str]] = None,
+            max_files: Optional[int] = None,
+        ):
+            # Stop if we have found max_count files
+            if max_files is not None and len(results) >= max_files:
+                return
+            listing = vfs.ls(uri)
+            for file_uri in listing:
+                # Avoid infinite recursion
+                if file_uri == uri:
+                    continue
+
+                if vfs.is_dir(file_uri):
+                    find(
+                        file_uri,
+                        include=include,
+                        exclude=exclude,
+                        suffixes=suffixes,
+                        max_files=max_files,
+                    )
+
+                else:
+                    match = match_uri(
+                        search_uri=search_uri,
+                        file_uri=file_uri,
+                        include=include,
+                        exclude=exclude,
+                        suffixes=suffixes,
+                    )
+                    if match:
+                        results.append(file_uri)
+
+        # Add one trailing slash to search_uri
+        search_uri = search_uri.rstrip("/") + "/"
+        results = []
+        find(
+            search_uri,
+            include=include,
+            exclude=exclude,
+            suffixes=suffixes,
+            max_files=max_files,
+        )
+        if max_files is not None:
+            return results[:max_files]
+        else:
+            return results
+
+
+def find_uris_tiledb_group(
+    search_uri: str,
+    config: Optional[Mapping[str, Any]] = None,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+    verbose: bool = False,
+) -> Sequence[str]:
+    with tiledb.scope_ctx(config):
+
+        def find(
+            uri: str,
+            path: str,
+            include: str = "*",
+            exclude: Sequence[str] = (),
+            suffixes: Optional[Sequence[str]] = None,
+            max_files: Optional[int] = None,
+        ):
+            # Stop if we have found max_count files
+            if max_files is not None and len(results) >= max_files:
+                return
+            obj_type = tiledb.object_type(uri)
+            if obj_type == "array":
+                match = match_uri(
```

Comment:
Question to make sure I understand the logic, is this matching requiring the array names to match expected file name?

Reply:
No this is matching the full path of the file inside the group structure.

We maintain the path as we follow the different groups in order to support similar matching functionality as the other directory based functions.